### ISSUE TITLE
Race-Condition Proof?

### ISSUE THREAD

**Issue by caseykickserv** [djreimer/sequenced#12]
Yes this is a hard problem to solve...However I have noted when performing imports or bulk model creation this library generates non-unique sequences.

```
class Person < ActiveRecord::Base
  validates_presence_of :account_id
   acts_as_sequenced scope: :account_id, column: :badge_number
end
```

In 2 consoles run the following command at the same time

```
 1000.times{ |i| Person.new(account_id: 1) }
```

I have a couple of strategies/ideas, but I thought it was worth asking. It is a problem that will come up in any bulk creation at scale or during concurrent processing which we do in the Rails world quite often through Resque or DJ or Sidekiq.

#### Strategies
* db locking
* In memory or in key-value store allocate the sequence and then use it. Similar to queue based strategies.
* Retry mechanism

I don't know how much of this is in the scope of a gem. But I wondered what we could do here to minimize this issue, or what is an approach that is used at scale that doesn't break sequence generation?

**Comment by ajb** [comment]
Hey @caseykickserv, I'd be curious to hear how you're approaching this issue. I've opened #13 as mostly an experiment, but we'll probably deploy it and see how it does, too.

**Comment by samphilipd** [comment]
The other pull request I saw has issues. It will work on Postgres but only by accident, it doesn't offer much extra help for MySQL at all.

I tried a lot of different ways to make this gem concurrent-safe including isolatable transactions and auto-retries. The best I found is a scoped unique index and not null constraint in combination with table locking.

See this PR for an example https://github.com/djreimer/sequenced/pull/16

It only works on Postgres but doesn't have any negative impact on other DB adapters so this shouldn't be an obstacle for merging.
